studio/app/optinist/core/edit_ROI/edit_ROI.py

```diff
@@ -14,125 +14,258 @@
+                "algo": cls.get_algo(filepath),
+                "file_path": filepath,
             },
         )
 
 
-class EditRoiUtils:
-    @classmethod
-    def conda(cls, config):
-        algo = config["algo"]
-        if "conda_name" in edit_roi_wrapper_dict[algo]:
-            conda_name = edit_roi_wrapper_dict[algo]["conda_name"]
-            return find_condaenv_filepath(conda_name) if conda_name else None
+class EditROI:
+    def __init__(self, file_path):
+        self.node_dirpath = os.path.dirname(file_path)
+        self.function_id = self.node_dirpath.split("/")[-1]
 
-        return None
+        self.output_info: Dict = PickleReader.read(self.pickle_file_path)
+        self.tmp_output_info: Dict = (
+            PickleReader.read(self.tmp_pickle_file_path)
+            if os.path.exists(self.tmp_pickle_file_path)
+            else {}
+        )
 
-    @classmethod
-    def execute(cls, config):
-        algo = config["algo"]
-        node_dirpath = config["node_dirpath"]
-        action = config["action"]
-        params = config["params"]
+        self.tmp_data: EditRoiData = self.tmp_output_info.get(
+            "edit_roi_data", self.output_info.get("edit_roi_data")
+        )
+        self.tmp_iscell = self.tmp_output_info.get(
+            "iscell", self.output_info.get("iscell")
+        ).data
+
+        print("start edit roi:", self.function_id)
+
+    @property
+    def pickle_file_path(self):
+        return glob(join_filepath([self.node_dirpath, "[!tmp_]*.pkl"]))[0]
 
-        func = copy.deepcopy(edit_roi_wrapper_dict[algo]["function"][action])
-        output_info = func(node_dirpath, **params)
-        del func
-        gc.collect()
+    @property
+    def tmp_pickle_file_path(self):
+        return join_filepath([self.node_dirpath, f"tmp_{self.function_id[:-11]}.pkl"])
 
-        workflow_dirpath = os.path.dirname(node_dirpath)
+    @property
+    def shape(self):
+        return self.tmp_data.im.shape[1:]
+
+    @property
+    def num_cell(self):
+        return self.tmp_data.im.shape[0]
+
+    def get_status(self) -> RoiStatus:
+        return self.tmp_data.status()
+
+    def add(self, roi_pos):
+        new_roi = create_ellipse_mask(self.shape, roi_pos)
+        new_roi = new_roi[np.newaxis, :, :] * self.num_cell
+
+        self.tmp_data.temp_add_roi[self.num_cell] = roi_pos
+        self.tmp_iscell = np.append(self.tmp_iscell, CellType.TEMP_ADD)
+        self.tmp_data.im = np.vstack((self.tmp_data.im, new_roi))
+
+        info = {
+            "cell_roi": RoiData(
+                np.nanmax(
+                    self.tmp_data.im[self.tmp_iscell != CellType.NON_ROI], axis=0
+                ),
+                output_dir=self.node_dirpath,
+                file_name="cell_roi",
+            ),
+            "iscell": IscellData(self.tmp_iscell),
+            "edit_roi_data": self.tmp_data,
+        }
+        self.__update_pickle_for_roi_edition(self.tmp_pickle_file_path, info)
```

Comment:
@quanpython 
`tmp_pickle_file` seems to have the same file size as the existing `pickle_file`,
is it necessary to duplicate the contents of `pickle_file`?
- I think that `tmp_pickle_file` should only hold the necessary information for ROI Edit.
- However, if it will simplify the process, I think it is OK to duplicate the pickle_file.

Reply:
@itutu-tienday
I have excluded unnecessary data from the `tmp_pickle_file`.

Reply:
@quanpython 
It seems that the pkl file has indeed been reduced, but after commit, the following system error (FE) seems to occur when selecting roi from the "Select Item" pull-down on the Visualize screen.
-> Are you sure you are not deleting even necessary data? (At commit time, don't you need to merge the difference information of roi edit with the original pkl?)

```
ERROR
invalid VisualaizeItemType
    at http://localhost:3001/static/js/bundle.js:33867:11
    at http://localhost:3001/static/js/bundle.js:425211:26
    at memoizedSelector (http://localhost:3001/static/js/bundle.js:453613:36)
    at getSnapshotWithSelector (http://localhost:3001/static/js/bundle.js:453659:20)
    at updateSyncExternalStore (http://localhost:3001/static/js/bundle.js:411221:26)
    at Object.useSyncExternalStore (http://localhost:3001/static/js/bundle.js:412168:18)
    at useSyncExternalStore (http://localhost:3001/static/js/bundle.js:434053:25)
    at useSyncExternalStoreWithSelector (http://localhost:3001/static/js/bundle.js:453668:19)
    at useSelector (http://localhost:3001/static/js/bundle.js:425257:27)
    at DisplayDataItemLayoutMenuIcon (http://localhost:3001/static/js/bundle.js:20522:79)
```

Reply:
@itutu-tienday
Sorry, the data needs to be added again when committing

Reply:
@quanpython 
I have confirmed that the tmp pkl file has been merged with the original in commit.
However, we have confirmed the following separate issues

- The error "invalid VisualaizeItemType" above seems not to have been resolved
  - If it looks like we need to work with FE, please do so.

- The following error log was recorded by BE during ROI confirmation for suite2p (suite2p_roi). Please confirm the contents.
  ```
  2023-11-28 12:26:03,981 INFO:     text_handler():503 - rule edit_ROI:
  ~~~~~~
    File "/home/user1/barebone-studio/studio/app/optinist/core/edit_ROI/edit_ROI.py", line 257, in __update_whole_nwb
      Runner.save_all_nwb(whole_nwb_path, last_output_info["nwbfile"])
    File "/home/user1/barebone-studio/studio/app/common/core/rules/runner.py", line 111, in save_all_nwb
      nwbconfig = merge_nwbfile(nwbconfig, x)
    File "/home/user1/barebone-studio/studio/app/optinist/core/nwb/nwb_creater.py", line 446, in merge_nwbfile
      old_nwbfile[pattern][function_id].update(
  AttributeError: 'list' object has no attribute 'update'
  ```

Reply:
@quanpython 
I will recheck the reproduction procedure.
*But since FE errors occur frequently in our environment, I will check after the FE errors are corrected.